File: Code/Python/UPM.py
# Data analysis of UPM tests
"""
Created on May 6 2019
@author: Tasdeeq Sofi
Data analysis written for the tests carried at UPM
"""
#Import different modules
import numpy as np
import pandas
#import scipy as sp
import scipy.io 
import matplotlib.pyplot as plt
import os # Directory maniplulater module 
import tkinter# ????
import tkinter.filedialog

#Preapre workspace      


#Load working directory
main_dir = tkinter.filedialog.askdirectory()
files = os.listdir(main_dir)
folders=[] # Initailize
nb_folders = 0 # find the number of folders
for f in files:
    filepath = os.path.join(main_dir, f)
    if os.path.isdir(filepath):
     nb_folders = nb_folders +1
     folders.append(f)

# Define parameters and Initialize
#legend_Var = zeros();
lgd1 = ['Free', 'S1', 'S2', 'S3', 'S4-Hammer', 'S4', 'S5-Hammer','S5']# check, python maybe loading the files in a different order
lgd2= ['Blutack','Hammer','Pristine']

# Load and plot all working files
for i in range (len(folders)):
    folder_name = folders[i]
    folderpath = os.path.join(main_dir,folder_name )
    work_files = os.listdir(folderpath)
    
    for i1 in range(len(work_files)):
        [filename, file_ext] = os.path.splitext(work_files[i1])
        filepath = os.path.join(folderpath,work_files[i1])
        if file_ext == '.csv':
            data = pandas.read_csv(filepath)
            Fr = data.iloc[:,0] # Frequency 
            data2 = data.iloc[:,1] # Capacitance and Impdance modulus
            data3 = data.iloc[:,2] # Resistance and angle of impedance vector          
        
        if folder_name == '01_Capacitance':
            fig1 = plt.figure(1)
#            fig, axs= plt.subplots(4,2)
#            fig.subplots_adjust(hspace = .5, wspace=.001)
#            axs[i1].plot(Fr, data2)
            Cp = data2
            plt.plot(Fr, Cp)
            plt.xlabel ('Frequency [Hz]')
            plt.ylabel ('Capacitance [Farad]')
            plt.legend (lgd1)
        
        if folder_name == '02_Impedance':
            phi=((np.math.pi)/180)*data3 #Convert into radians
            ZR = data2*np.cos(phi) # Real Impedance
            ZI = data2*np.sin(phi)# Imaginary Impedance
            AI = abs(1./ZI) # Imaginary Admittance
            fig2 = plt.figure(2)
            plt.plot(Fr, ZR)
            plt.xlabel ('Frequency [Hz]')
            plt.ylabel ('Impedance [Ohm]')
            plt.legend (lgd1)
            fig3 = plt.figure(3)
            plt.plot(Fr, AI)
            plt.xlabel ('Frequency [Hz]')
            plt.ylabel ('Admittance [Seimens]')
            plt.legend (lgd1)
            
        if folder_name == '03_GuidedWave':
            mat = scipy.io.loadmat(filepath)
            Ac_signal = mat['A']
            Rv_signal = mat['B']
            fig4 = plt.figure(4)
            plt.plot(Ac_signal)
            plt.xlabel ('Time [S]')
            plt.ylabel ('Amplitude [v]')
            plt.legend (lgd2)
            fig5 = plt.figure(5)
            plt.plot(Rv_signal)
            plt.xlabel ('Time [S]')
            plt.ylabel ('Amplitude [v]')
            plt.legend (lgd2)
            
        
    
    
    
    
    
    
    
    
    
    
    
    
    
    

        

File: Code/Matlab/UPM.m
%% Author: Tasdeeq Sofi, Project: ENHNANCE H2020
% Date: 23/03/21
% Code written for analyzing the tests carried out at UPM
% Capacitance, Impedance and Guided wave Measurements were carried out for DuraAct sensors secondary bonded to PEEK panels
% Instructions: Run the folder containing the capacitance, impedance and the guide wave measurements in seprate folders "e.g., 01_TestData located at UPM_22-03-21_CP-IP-GW"

%% Prepare workspace and Define parameters if any
clear;
close all;
clc;
warning off;

%% Load working directory and Intialize
% Input the folder to work with
main_dir=uigetdir('C:\Users\tsofi\OneDrive\01_ENHAnCE\07_ExperimentalData'); 
folders = dir(main_dir);
pathstr = folders(1).folder;

% Checks the number of folders to process
nb_folders=0;
for j=1:length (folders)
    if (folders(j).isdir && ~strcmp(folders(j).name,'.') && ~strcmp(folders(j).name,'..'))
        nb_folders=nb_folders+1;
    end
end

% Initalize
time = zeros();
% legend_Var = zeros();
f1=figure(1);
f2=figure(2);
f3=figure(3);
f4=figure(4);
f5=figure(5);
%% Load csv files uisng for loop, obatin frequncy analysis graphs
for i = 1:length(folders) 
   
   if (folders(i).isdir) % Check if it is a folder
       if ( ~strcmp(folders(i).name,'.') && ~strcmp(folders(i).name,'..'))% because of '.' and '..'
           % Load all working folders
           fodler_name = folders(i).name;
           folderpath = fullfile(pathstr,fodler_name);
           files = dir(folderpath);
           % Load and plot all csv files
           for i1 = 3:length(files) % because of '.' and '..'
               file_name = files(i1).name;
               filename = fullfile(folderpath, file_name); 
               [folderpath,name,ext] = fileparts(filename);
               if strcmp (ext, '.csv')==1
                  data = readtable(filename);
                  % Obtain frquency, Capacitance and Impedance       
                   Fr = data.(1); % Frequency 
                   data2 = data.(2); % Capacitance and Impdance modulus
                   data3 = data.(3); % Resistance and angle of impedance vector

                   if strcmp (fodler_name, '01_Capacitance')==1    
                       % plot       
                       set(0, 'CurrentFigure', f1)
                       subplot(4,2,(i1-2));
                       Cp = data2;
                       plot (Fr,Cp,'LineWidth',1.5)
                       xlabel ('Frequency [Hz]','FontSize', 14)
                       ylabel ('Capacitance [F]','FontSize', 14)
                       title (file_name(1:(end-8))) 
                       set(0, 'CurrentFigure', f2)
                       plot (Fr,Cp,'LineWidth',1.5)
                       hold on
                       xlabel ('Frequency [Hz]','FontSize', 14)
                       ylabel ('Capacitance [F]','FontSize', 14)
                       legend ('Free', 'S1', 'S2', 'S3', 'S4-Hammer', 'S4', 'S5-Hammer','S5', 'Location','southeast'); 
                   elseif strcmp (fodler_name, '02_Impedance')==1
                       % Obtain frquency, Impedance modulus and phase-angle
                       phi=(pi/180).*data3;% Convert into radians
                       Zr_r = data2.*cos(phi);%  Real Impedance
                       Zr_i = data2.*sin(phi);%  Imaginary Impedance
                       Ad_r = abs(1./Zr_r);%  Real Admittance
                       Ad_i = abs(1./Zr_i);%  Imaginary Admittance
                       % plot
                       set(0, 'CurrentFigure', f3)
                       subplot(4,2,(i1-2));
                       plot (Fr,Ad_i,'LineWidth',1.5)
                       xlabel ('Frequency [Hz]','FontSize', 14)
                       ylabel ('Imaginary Admittance [S]','FontSize', 14)
                       title (file_name(1:(end-8)))                   
                       set(0, 'CurrentFigure', f4)
                       plot (Fr,Ad_i,'LineWidth',1.5)
                       hold on
                       xlabel ('Frequency [Hz]','FontSize', 14)
                       ylabel ('Imaginary Admittance [S]','FontSize', 14)
                       legend ('Free', 'S1', 'S2', 'S3', 'S4-Hammer', 'S4', 'S5-Hammer','S5', 'Location','northwest');  
                       set(0, 'CurrentFigure', f5)
                       plot (Fr,Zr_r,'LineWidth',1.5)
                       hold on
                       xlabel ('Frequency [Hz]','FontSize', 14)
                       ylabel ('Real impedance [Ohm]','FontSize', 14)                       
            %            c = file_name(1:(end-4));
            %            legend_Var(i,1)=c;
            %            legend (legend_Var(1,1),legend_Var(2),legend_Var(3),legend_Var(4),legend_Var(5),legend_Var(6),legend_Var(7),legend_Var(8));
                       legend ('Free', 'S1', 'S2', 'S3', 'S4-Hammer', 'S4', 'S5-Hammer','S5', 'Location','northeast');         
                   end
               elseif  strcmp (fodler_name, '03_GuidedWave')==1          
                   data = load(filename); % '.mat' files
                   Ad_r = data.A;
                   B = data.B;
                   for i2 =1:length(Ad_r)% obtain time vector
                        time(1,1) = data.Tstart;
                        time((i2+1),1) = data.Tinterval + time(i2,1);
                   end
                   time = time(1:length(Ad_r));
                   figure(6);
                   plot (time,Ad_r,'LineWidth',1)
                   hold on
                   legend ('Blutack','Hammer','Pristine');                 
                   figure(7);
                   plot (time,B,'LineWidth',1)
                   hold on
                   legend ('Blutack','Hammer','Pristine');
                   title (file_name)
               else
%                    disp ('No working files found in the folder')
               end


           end
       end
   end

   
end